### ISSUE TITLE
how to enable subscriptions

### ISSUE THREAD

**Issue by barbalex** [mobxjs/mst-gql#247]
I have read:
- [hasura documentation on subscriptions](https://hasura.io/docs/1.0/graphql/manual/subscriptions/index.html) as that is my graphql api
- [the very short documentation I found](https://github.com/mobxjs/mst-gql#subscribequery-variables-ondata---void) in mst-gql itself. It basically only says: "The gqlWsClient needs to be set during the store creation to make this possible"
- [the readme for graphql-request](https://github.com/prisma-labs/graphql-request). Which baffles me [as graphql-request does not currently support subscriptions](https://github.com/prisma-labs/graphql-request/issues/77)

So basically I do not know how this should work as it is not supported by graphql-request. Yet "The gqlWsClient needs to be set during the store creation to make this possible" seems to imply that it is possible.

I have also searched for an analog method to `createHttpClient`, something like `createWsClient`. But it does not seem to exist.

So my questions are:
- can it really be done?
- if yes: how?

**Comment by chrisdrackett** [comment]
that would be great!

**Comment by barbalex** [comment]
Additional caveat:

When using server side rendered tools like gatsby/next/nuxt it is necessary to prevent using subscriptions server side. Otherwise there will be an error because the server is missing a websocket implementation.

This is how I do it in gatsby:
```js
const gqlHttpClient = createHttpClient(constants.graphQlUri)
const tokenWithRoles =
  typeof window !== 'undefined'
    ? window.localStorage.getItem('token') || 'none'
    : 'none'
gqlHttpClient.setHeaders({ authorization: `Bearer ${tokenWithRoles}` })

// ws client only works in the browser
// need to prevent gatsby from executing it server side
// see: https://github.com/apollographql/subscriptions-transport-ws/issues/333#issuecomment-359261024
let gqlWsClient
let storeOptions = {
  gqlHttpClient,
}
if (typeof window !== 'undefined') {
  // https://www.npmjs.com/package/subscriptions-transport-ws#hybrid-websocket-transport
  gqlWsClient = new SubscriptionClient(constants.graphQlWsUri, {
    reconnect: true,
    connectionParams: {
      headers: { authorization: `Bearer ${tokenWithRoles}` },
    },
  })
  // https://github.com/mobxjs/mst-gql/blob/master/src/MSTGQLStore.ts#L42-L43
  storeOptions = {
    gqlHttpClient,
    gqlWsClient,
  }
}

const store = RootStore.create(undefined, storeOptions)
```
